shadow-utils-4.1.4.2/man/groups.1
'\" t
.\"     Title: groups
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.75.1 <http://docbook.sf.net/>
.\"      Date: 07/24/2009
.\"    Manual: User Commands
.\"    Source: User Commands
.\"  Language: English
.\"
.TH "GROUPS" "1" "07/24/2009" "User Commands" "User Commands"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
groups \- display current group names
.SH "SYNOPSIS"
.HP \w'\fBgroups\fR\ 'u
\fBgroups\fR [\fIuser\fR]
.SH "DESCRIPTION"
.PP
The
\fBgroups\fR
command displays the current group names or ID values\&. If the value does not have a corresponding entry in
/etc/group, the value will be displayed as the numerical group value\&. The optional
\fIuser\fR
parameter will display the groups for the named
\fIuser\fR\&.
.SH "NOTE"
.PP
Systems which do not support concurrent group sets will have the information from
/etc/group
reported\&. The user must use
\fBnewgrp\fR
or
\fBsg\fR
to change their current real and effective group ID\&.
.SH "FILES"
.PP
/etc/group
.RS 4
Group account information\&.
.RE
.SH "SEE ALSO"
.PP

\fBnewgrp\fR(1),
\fBgetgid\fR(2),
\fBgetgroups\fR(2),
\fBgetuid\fR(2)\&.


gmime-38b35d8/ChangeLog
2015-01-23  Sunghyun Kwon     <sh0701.kwon@samsung.com>

    * gmime/gmime-gpg-context.c (gpg_ctx_set_mode): Added the condition
    (GPG_CTX_MODE_SIGN_ENCRYPT) 

2013-02-19  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.15

2013-02-19  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-filter-html.c (html_convert): Fixed to convert the
	last line of input even if it doesn't end with a \n.

2013-02-18  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.14

2013-02-18  Jeffrey Stedfast  <fejj@gnome.org>

	* util/url-scanner.c (url_addrspec_start): Fixed the logic to
	actually scan for the beginning of the mailbox.

2012-12-09  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.13

2012-11-26  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-header.c (g_mime_header_iter_first): Make sure the
	list isn't empty.
	(g_mime_header_iter_last): Same.

2012-11-23  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-header.c (g_mime_header_list_prepend): Prepend the
	header to the list, don't *append* it. Oops.
	(g_mime_header_list_clear): Don't forget to reset the linked list
	so that it doesn't contained free'd nodes.

2012-11-02  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.12

2012-10-27  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-stream-filter.c (stream_read): Initialize presize to
	READ_PAD instead of READ_SIZE to prevent potential buffer
	underruns.

2012-10-20  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-filter-gzip.c (gunzip_filter): Apparently
	Z_BUF_ERROR isn't actually an error unless there is input
	available. Thanks to Pavlo <devgs@ukr.net> for this patch.

2012-10-19  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-parser.c (parser_push_boundary): Don't use strcmp to
	check if the boundary is an mbox boundary, use a direct pointer
	comparison so that we can handle Content-Type boundary parameters
	with a value of "From "

2012-10-19  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-parser.c (parser_step_headers): Don't assume that
	the first non-header line we find indicates the start of the
	content unless we've already gotten some basic headers.

	Fix for bug #686223.

2012-10-18  Jeffrey Stedfast  <jeff@xamarin.com>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.11

2012-09-12  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-gpg-context.c (gpg_ctx_parse_status): Unref the
	certificate after adding it to the list (adding it to the list
	refs it).

	* gmime/gmime-multipart-encrypted.c
	(g_mime_multipart_encrypted_decrypt): If the 'result' out param is
	null, unref the GMimeDecryptResult to avoid leaking.

2012-08-13  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-utils.c (tokenize_rfc2047_phrase): Don't forget to
	initialize 'ascii' to TRUE.

2012-08-05  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-stream-pipe.c (stream_flush): Pipes and sockets
	cannot be fsync()'d, so don't even try!

2012-08-05  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-utils.c (rfc2047_token_new_encoded_word): Make sure
	to properly handle broken encoded-word tokens in the form:
	=?charset?q?=.

	Fixes the g_malloc abort() in bug #679388

2012-08-05  Jeffrey Stedfast  <jeff@xamarin.com>

	* gmime/gmime-multipart-signed.c (check_protocol_supported): Oops,
	need to compare against xsupported in the second string compare.

	Fixes bug #674032

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.10

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/Makefile.am: Always build gmime-gpg-context.[c,h] and
	gmime-pkcs7-context.[c,h]

	* gmime/gmime-gpg-context.c: Armor internals with #ifdef
	ENABLE_CRYPTOGRAPHY so that if crypto is disabled, only the
	skeletal API is built.

	* gmime/gmime.h: Include gmime-pkcs7-context.h

	Fixes bug #676692.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-utils.c (rfc2047_encode_get_rfc822_words): Use
	is_blank() instead of is_lwsp() so that we only treat space and
	tab as word delimiters (we want to rfc2047 encode \r's and
	\n's). Check for ctrl characters when deciding what the type and
	encoding of the word is.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-utils.c (rfc2047_encode_get_rfc822_words): When
	breaking up really long atom tokens, change the word type to
	RFC2047 so that we encode them, thus preventing forced header
	folding which cannot be unmunged.
	(header_fold): Fixed for cases where the first word in a header is
	68 chars. (Note: this fix is largely unnecessary due to the above fix)

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-message.c: When setting a header, make sure to
	always clear the header list cache of the toplevel mime_part as
	well.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-param.c (rfc2184_param_new): Only decode the charset
	and lang if the parameter is actually encoded. Thanks to Peter
	Bloomfield for this patch.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-gpg-context.c (gpg_ctx_parse_status): Keep track of
	DECRYPTION_OKAY status.
	(gpg_decrypt): Only return NULL if decryption failed (e.g. we
	didn't get a DECRYPTION_OKAY status).

	Fixes bug #677088.

2012-06-13  Jeffrey Stedfast  <fejj@gnome.org>

	* docs/reference/building.sgml: Updated docs.

	Fixes bug #676341

2012-04-22  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.9

2012-04-22  Jeffrey Stedfast  <fejj@gnome.org>

	* gmime/gmime-multipart-signed.c (g_mime_multipart_signed_verify):
	Match protocols like application/x-pkcs7-signature, too.

	Fixes bug #674032.

2012-04-19  Jeffrey Stedfast  <fejj@gnome.org>

	* README: Bumped version

	* configure.ac: Bumped version to 2.6.8


shadow-utils-4.1.4.2/src/groups.c
/*
 * Copyright (c) 1991 - 1993, Julianne Frances Haugh
 * Copyright (c) 1996 - 2000, Marek Michałkiewicz
 * Copyright (c) 2001 - 2006, Tomasz Kłoczko
 * Copyright (c) 2007 - 2008, Nicolas François
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the copyright holders or contributors may not be used to
 *    endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT
 * HOLDERS OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <config.h>

#ident "$Id: groups.c 2849 2009-04-30 21:08:49Z nekral-guest $"

#include <grp.h>
#include <pwd.h>
#include <stdio.h>
#include "defines.h"
#include "prototypes.h"
/*
 * Global variables
 */
char *Prog;

/* local function prototypes */
static void print_groups (const char *member);

/*
 * print_groups - print the groups which the named user is a member of
 *
 *	print_groups() scans the groups file for the list of groups which
 *	the user is listed as being a member of.
 */
static void print_groups (const char *member)
{
	int groups = 0;
	struct group *grp;
	struct passwd *pwd;
	bool flag = false;

	pwd = getpwnam (member); /* local, no need for xgetpwnam */
	if (NULL == pwd) {
		(void) fprintf (stderr, _("%s: unknown user %s\n"),
		                Prog, member);
		exit (EXIT_FAILURE);
	}

	setgrent ();
	while ((grp = getgrent ()) != NULL) {
		if (is_on_list (grp->gr_mem, member)) {
			if (0 != groups) {
				(void) putchar (' ');
			}
			groups++;

			(void) printf ("%s", grp->gr_name);
			if (grp->gr_gid == pwd->pw_gid) {
				flag = true;
			}
		}
	}
	endgrent ();

	/* The user may not be in the list of members of its primary group */
	if (!flag) {
		grp = getgrgid (pwd->pw_gid); /* local, no need for xgetgrgid */
		if (NULL != grp) {
			if (0 != groups) {
				(void) putchar (' ');
			}
			groups++;

			(void) printf ("%s", grp->gr_name);
		}
	}

	if (0 != groups) {
		(void) putchar ('\n');
	}
}

/*
 * groups - print out the groups a process is a member of
 */
int main (int argc, char **argv)
{
#ifdef HAVE_GETGROUPS
	long sys_ngroups;
	int ngroups;
	GETGROUPS_T *groups;
	int pri_grp; /* TODO: should be GETGROUPS_T */
	int i;
#else
	char *logname;
	char *getlogin ();
#endif

#ifdef HAVE_GETGROUPS
	sys_ngroups = sysconf (_SC_NGROUPS_MAX);
	groups = (GETGROUPS_T *) malloc (sizeof (GETGROUPS_T) * sys_ngroups);
#endif
	(void) setlocale (LC_ALL, "");
	(void) bindtextdomain (PACKAGE, LOCALEDIR);
	(void) textdomain (PACKAGE);

	/*
	 * Get the program name so that error messages can use it.
	 */
	Prog = Basename (argv[0]);

	if (argc == 1) {

		/*
		 * Called with no arguments - give the group set for the
		 * current user.
		 */

#ifdef HAVE_GETGROUPS
		/*
		 * This system supports concurrent group sets, so I can ask
		 * the system to tell me which groups are currently set for
		 * this process.
		 */
		ngroups = getgroups (sys_ngroups, groups);
		if (ngroups < 0) {
			perror ("getgroups");
			exit (EXIT_FAILURE);
		}

		/*
		 * The groupset includes the primary group as well.
		 */
		pri_grp = getegid ();
		for (i = 0; i < ngroups; i++) {
			if (pri_grp == (int) groups[i]) {
				break;
			}
		}

		if (i != ngroups) {
			pri_grp = -1;
		}

		/*
		 * Print out the name of every group in the current group
		 * set. Unknown groups are printed as their decimal group ID
		 * values.
		 */
		if (-1 != pri_grp) {
			struct group *gr;
			/* local, no need for xgetgrgid */
			gr = getgrgid (pri_grp);
			if (NULL != gr) {
				(void) printf ("%s", gr->gr_name);
			} else {
				(void) printf ("%d", pri_grp);
			}
		}

		for (i = 0; i < ngroups; i++) {
			struct group *gr;
			if ((0 != i) || (-1 != pri_grp)) {
				(void) putchar (' ');
			}

			/* local, no need for xgetgrgid */
			gr = getgrgid (groups[i]);
			if (NULL != gr) {
				(void) printf ("%s", gr->gr_name);
			} else {
				(void) printf ("%ld", (long) groups[i]);
			}
		}
		(void) putchar ('\n');
#else
		/*
		 * This system does not have the getgroups() system call, so
		 * I must check the groups file directly.
		 */
		logname = getlogin ();
		if (NULL != logname) {
			print_groups (logname);
		} else {
			exit (EXIT_FAILURE);
		}
#endif
	} else {

		/*
		 * The invoker wanted to know about some other user. Use
		 * that name to look up the groups instead.
		 */
		print_groups (argv[1]);
	}
	return EXIT_SUCCESS;
}

